### PR TITLE
Chill 0.3 with rose pkg

### PR DESCRIPTION
**Pull request by dhuth**

### CODE REVIEW

**Comment by dhuth** [comment]
Updating CHiLL package and adding rose-for-chill dependency.

**Comment by dhuth** [comment]
Changes to the rose package can be discarded.

File: var/spack/repos/builtin/packages/chill/package.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+##############################################################################
+# Copyright (c) 2013-2019, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
```
Comment (baberlevi) [inline var/spack/repos/builtin/packages/chill/package.py]: This seems to be the old boilerplate from pre-relicense times

File: var/spack/repos/builtin/packages/chill/package.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#-----------------------------------------------------------------------------
+# Author: Derick Huth <derick.huth@utah.edu>
+#-----------------------------------------------------------------------------
+from spack import *
+
+
+class Chill(Package):
+    """FIXME: Put a proper description of your package here."""
```
Comment (baberlevi) [inline var/spack/repos/builtin/packages/chill/package.py]: fixme needs replaced with the package description

File: var/spack/repos/builtin/packages/chill/package.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    homepage = "http://github.com/CtopCsUtahEdu"
+    url      = "https://github.com/CtopCsUtahEdu/chill/archive/v0.3.tar.gz"
+    git      = "https://github.com/CtopCsUtahEdu/chill.git"
+
+    #version('0.3', sha256='574b622368a6bfaadbe9c1fa02fabefdc6c006069246f67d299f943b7e1d8aa3')
```
Comment (baberlevi) [inline var/spack/repos/builtin/packages/chill/package.py]: It looks like you've got a release - what's the reason for commenting out ?

File: var/spack/repos/builtin/packages/chill/package.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#-----------------------------------------------------------------------------
+# Author: Derick Huth <derick.huth@utah.edu>
+#-----------------------------------------------------------------------------
+from spack import *
+
+
+class Chill(Package):
```
Comment (baberlevi) [inline var/spack/repos/builtin/packages/chill/package.py]: could this be changed to an `AutotoolsPackage` instead of just package ? Looks like you support Autotools or cmake, right ?

File: var/spack/repos/builtin/packages/iegenlib/package.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+# Author: Derick Huth <derick.huth@utah.edu>
+#------------------------------------------------------------------------------
+#   Note: This package was created by the ctop reaserch group at
+#         the University of Utah, Shool of Computing and is unifiliated
+#         with IEGenLib
+# -----------------------------------------------------------------------------
+from spack import *
+
+
+class Iegenlib(Package):
```
Comment (baberlevi) [inline var/spack/repos/builtin/packages/iegenlib/package.py]: same comment as above about build system type

File: var/spack/repos/builtin/packages/iegenlib/package.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+# Copyright (c) 2013-2018, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/spack/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
```
Comment (baberlevi) [inline var/spack/repos/builtin/packages/iegenlib/package.py]: same as above - old boilerplate

File: var/spack/repos/builtin/packages/chill/package.py (hunk 4 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/chill/package.py]: If I remember right, there were some additional tasks that didn't fit into the AutotoolsPackage workflow.

Comment (citibeth) [reply var/spack/repos/builtin/packages/chill/package.py]: It should be possible to extend the `AutotoolsPackage` workflow.

Comment (dhuth) [reply var/spack/repos/builtin/packages/chill/package.py]: I'll take another look at it

**Comment by dhuth** [comment]
Is one of the travis-ci tests timing out?

**Comment by baberlevi** [comment]
Yes, looks like it did. I retriggered the step that timed out

**Comment by dhuth** [comment]
Looks like it's still timing out. Is there a way to find out whats causing that?

**Comment by dhuth** [comment]
This is still hanging. Is there a way for me to restart it or diagnose the problem?

**Comment by scheibelp** [comment]
> This is still hanging. Is there a way for me to restart it or diagnose the problem?

Closing/reopening will restart it. That's what I recommend when there is a timeout. When I notice that I can restart the tests without doing that (which I did here).

**Comment by dhuth** [comment]
Is there a way to see what went wrong with codecov/project/build_systems?

File: var/spack/repos/builtin/packages/chill/package.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    depends_on('rose@0.9.10.0 +cxx11 ^boost@1.66.0 cxxstd=11')
+    depends_on('iegenlib')
+    depends_on('isl')
+    depends_on('python')
+
+    @run_before('configure')
+    def bootstrap(self, spec, prefix):
+        bash = wich('bash')
+        bash('./bootstrap')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/chill/package.py]: Is this standard for Autotools?  Why does this not look like the typical "autorebuild" stuff?

File: var/spack/repos/builtin/packages/chill/package.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+    def setup_environment(self, spack_env, run_env):
+        rose_home = self.spec['rose'].prefix
+        boost_home = self.spec['boost'].prefix
+        iegen_home = self.spec['iegenlib'].prefix
+
+        spack_env.append_path('LD_LIBRARY_PATH', rose_home + '/lib')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/chill/package.py]: Just use `self.spec['rose'].prefix.lib`, etc.

File: var/spack/repos/builtin/packages/iegenlib/package.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# -----------------------------------------------------------------------------
+# Author: Derick Huth <derick.huth@utah.edu>
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/iegenlib/package.py]: Can you add yourself in the `maintainers` variable inside the package?  Grep for examples...

File: var/spack/repos/builtin/packages/iegenlib/package.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+# -----------------------------------------------------------------------------
+#   Note: This package was created by the ctop reaserch group at
+#         the University of Utah, Shool of Computing and is unifiliated
+#         with IEGenLib
+# -----------------------------------------------------------------------------
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/iegenlib/package.py]: Is this message needed?  It's not typical for Spack.  The Git Repo and Spack history already shows who created it; but over time, it's likely that others might add to / change the package significantly.

File: var/spack/repos/builtin/packages/iegenlib/package.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class Iegenlib(AutotoolsPackage):
+    """Inspector/Executor Generation Library for manipulating sets
+       and relations with uninterpreted function symbols. """
+
+    homepage = "http://github.com/CompOpt4Apps"
+    git      = "https://github.com/CompOpt4Apps/IEGenLib.git"
+
+    version('master', branch='master')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/iegenlib/package.py]: Do you have any checksummable (non-changing) versions you can add?  That's important for security (see past discussions for why).  If there isn't at least one release of this package, maybe it's not ready for Spack yet...

File: var/spack/repos/builtin/packages/iegenlib/package.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/iegenlib/package.py]: Why doesn't `autoreconf` need to be run, as is typical when checking Autotools packages out of a git repo?  Note that checking the `./configure` script into the Git repo is a bad idea (search around for why).

File: var/spack/repos/builtin/packages/iegenlib/package.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    depends_on('cmake@2.6:', type='build')
+    depends_on('isl')
+    depends_on('texinfo', type='build')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/iegenlib/package.py]: I assume this is for docs.  Can it be made optional?  Often, docs introduce a lot of dependencies that 90% of people don't want/need to build (because they get the docs already built off a website).

File: var/spack/repos/builtin/packages/rose/package.py [hunk 14]
```diff
@@ -4,79 +4,288 @@
...
 
-    variant('z3', default=False, description='Enable z3 theorem prover')
-    depends_on('z3', when='+z3')
+    variant("c", default=True, description="Enable c language support")
+    variant("cxx", default=True, description="Enable c++ language support")
 
-    build_directory = 'spack-build'
+    # Use spack install cxxflags=-std=c++11
+    variant("cxx11", default=False, description="Enable c++11 language support")
+
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/rose/package.py]: Remove these extra blank lines?

File: var/spack/repos/builtin/packages/rose/package.py [hunk 15]
```diff
@@ -4,79 +4,288 @@
...
-            'binaries' if '+binanalysis' in spec else '',
-            'c' if '+c' in spec else '',
-            'c++' if '+cxx' in spec else '',
-            'java' if '+java' in spec else '',
-            'fortran' if '+fortran' in spec else ''
+            "binaries" if "+binanalysis" in spec else "",
+            "c" if "+c" in spec else "",
+            "c++" if "+cxx" in spec else "",
+            "java" if "+java" in spec else "",
+            "fortran" if "+fortran" in spec else "",
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/rose/package.py]: Why is this change here?

File: var/spack/repos/builtin/packages/rose/package.py [hunk 16]
```diff
@@ -4,79 +4,288 @@
...
-    install_targets = ["install-core"]
+        if "+z3" in spec:
+            args.append("--with-z3={0}".format(spec["z3"].prefix))
+        else:
+            args.append("--without-z3")
+
+        if "+tests" in spec:
+            args.append("--enable-tests-directory")
+        else:
+            args.append("--disable-tests-directory")
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/rose/package.py]: Or try something like:
```
args.append("--%sable-tests-directory" % "en" if "+tests" in spec else "dis")
```
This would reduce the number of lines in this section by a factor of 5.  You can even do:
```
args = [
    "--%sable-tests-directory" % "en" if "+tests" in spec else "dis",
    "--%sable-java" % "en" if "+java" in spec else "dis",
    ...
]
```

File: var/spack/repos/builtin/packages/chill/package.py (hunk 7 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/chill/package.py]: This is a non-standard script that runs autoreconf and also generates test cases.

File: var/spack/repos/builtin/packages/iegenlib/package.py (hunk 10 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/iegenlib/package.py]: Will remove it then.

File: var/spack/repos/builtin/packages/chill/package.py (hunk 7 above)
Comment (citibeth) [reply var/spack/repos/builtin/packages/chill/package.py]: OK.. then you probably need this in your package:
```
depends_on('autoconf', type='build')
depends_on('automake', type='build')
depends_on('libtool',  type='build')
depends_on('m4',       type='build')
```

https://spack.readthedocs.io/en/latest/build_systems/autotoolspackage.html

File: var/spack/repos/builtin/packages/iegenlib/package.py (hunk 12 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/iegenlib/package.py]: IEGenLib is a cmake package with it's own ./confgiure script. It otherwise works like an autotools package.

File: var/spack/repos/builtin/packages/iegenlib/package.py (hunk 13 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/iegenlib/package.py]: I haven't been able to build it without texinfo.

File: var/spack/repos/builtin/packages/rose/package.py (hunk 14 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/rose/package.py]: This change is from an older version of the rose package to the one that is currently in the repository.

File: var/spack/repos/builtin/packages/chill/package.py [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        run_env.append_path('LD_LIBRARY_PATH', rose_home.lib)
+        run_env.append_path('LD_LIBRARY_PATH', boost_home.lib)
+        run_env.append_path('LD_LIBRARY_PATH', iegen_home.lib)
+
+        spack_env.set('ROSEHOME', rose_home)
+        spack_env.set('BOOSTHOME', boost_home)
+        spack_env.set('IEGENHOME', iegen_home)
+
+    def configure_args(self):
+        args = ['--with-rose={0}'.format(self.spec['rose'].prefix),
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/chill/package.py]: Should this be `rose_home`?

File: var/spack/repos/builtin/packages/chill/package.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Chill(AutotoolsPackage):
+    """A polyheadral compiler for autotuning"""
+
+    homepage = "http://github.com/CtopCsUtahEdu"
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/chill/package.py]: This should be `https://github.com/CtopCsUtahEdu/chill`

File: var/spack/repos/builtin/packages/chill/package.py [hunk 19]
```diff
@@ -0,0 +1,55 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Chill(AutotoolsPackage):
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/chill/package.py]: Why does this use Autotools?  Why not the CMake build for Chill?  Almost certainly, the Autotools build is / will be deprecated.  It would be better if Spack didn't go down that route and have to change later.

```
CHill can be built from two build systems, CMake or automake. (CMake is recomended for use with CLion)
```

File: var/spack/repos/builtin/packages/iegenlib/package.py (hunk 13 above)
Comment (citibeth) [reply var/spack/repos/builtin/packages/iegenlib/package.py]: Really?  I can't find any texinfo in the Autotools build.  And the CMake build seems to work fine if `texi2pdf` is not available.  See:

https://github.com/CtopCsUtahEdu/chill/search?q=chill.texi&unscoped_q=chill.texi

File: var/spack/repos/builtin/packages/chill/package.py (hunk 17 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/chill/package.py]: It could be. I have a bit of an aversion to adding unnecessary members to objects, but for the sake of clarity, it makes sense. Is setup_environment always called before configure_args?

File: var/spack/repos/builtin/packages/chill/package.py (hunk 19 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/chill/package.py]: As of right now, the autotools build is the one that's been most thoroughly tested.

File: var/spack/repos/builtin/packages/iegenlib/package.py (hunk 13 above)
Comment (dhuth) [reply var/spack/repos/builtin/packages/iegenlib/package.py]: IEGenLib specifically

**Comment by dhuth** [comment]
How long should I wait before I get worried about travis-ci?

**Comment by scheibelp** [comment]
Depending on the load it can take an hour or two to start. If there is a conflict (as there is here) then it may not start at all (so you would want to merge from develop and then push those updates here).

**Comment by dhuth** [comment]
I'm having a hard time interpreting the error from https://travis-ci.org/spack/spack/jobs/551020778

**Comment by scheibelp** [comment]
> I'm having a hard time interpreting the error from https://travis-ci.org/spack/spack/jobs/551020778

That's odd: the build that "matters" appears to be passing (https://travis-ci.org/spack/spack/builds/551428267?utm_source=github_status&utm_medium=notification) so I'm not sure where that came from.

I'm going to check this a bit more but generally LGTM

**Comment by scheibelp** [comment]
For some reason the github UI is displaying changes from https://github.com/spack/spack/commit/a49ce3fc321d2882bc3a8ad67042a43da5f5fdcf#diff-a1678e1b87b8717195c375013288d7c0 to the rose package. This does not appear to make any changes to Rose.

@dhuth I'm curious what merge/rebasing commands you were using on this

**Comment by dhuth** [comment]
That was a manual merge that should have been a merge from develop.
